src/components/functional_module/src/plugin_manager.cc

```diff
@@ -0,0 +1,410 @@
+      msg->protocol_version()) {
+    std::string msg_method;
+    // Request or notification from HMI
+    if (value.isMember("method") && value["method"].isString()) {
+      msg_method = value["method"].asCString();
+      // Response from HMI
+    } else if (value.isMember("result") && value["result"].isMember("method") &&
+               value["result"]["method"].isString()) {
+      msg_method = value["result"]["method"].asCString();
+      // Error response from HMI
+    } else if (value.isMember("error") && value["error"].isMember("data") &&
+               value["error"]["data"].isMember("method") &&
+               value["error"]["data"]["method"].isString()) {
+      msg_method = value["error"]["data"]["method"].asCString();
+    } else {
+      LOG4CXX_WARN(logger_,
+                   "Message with HMI protocol version can not be handled by "
+                   "plugin manager, because required 'method' field was not "
+                   "found, or was containing an invalid string.");
+      return false;
+    }
+
+    return (hmi_subscribers_.find(msg_method) != hmi_subscribers_.end());
+  }
+
+  return false;
+}
+
+void PluginManager::OnServiceStateChanged(ServiceState state) {
+  for (PluginsIterator it = plugins_.begin(); plugins_.end() != it; ++it) {
+    it->second->OnServiceStateChanged(state);
+  }
+}
+
+void PluginManager::OnHMIResponse(application_manager::MessagePtr msg) {
+  // TODO(PV)
+}
+
+void PluginManager::OnError(ModuleObserver::Errors error, ModuleID module_id) {
+  std::string error_string;
+  switch (error) {
+    case ModuleObserver::Errors::OUT_OF_MEMORY:
+      error_string = "Module run out of memory.";
+      break;
+    case ModuleObserver::Errors::FS_FAILURE:
+      error_string = "Plugin failed to run file system operation.";
+      break;
+    default:
+      break;
+  }
+  LOG4CXX_ERROR(logger_,
+                "Error " << error_string << " was received from module "
+                         << module_id);
+  // TODO(PV)
+}
+
+void PluginManager::RemoveAppExtension(uint32_t app_id) {
+  for (PluginsIterator it = plugins_.begin(); plugins_.end() != it; ++it) {
+    it->second->RemoveAppExtension(app_id);
+  }
+}
+
+bool PluginManager::IsAppForPlugins(
+    application_manager::ApplicationSharedPtr app) {
+  DCHECK(app);
+  if (!app) {
+    return false;
+  }
+
+  bool res = false;
+  for (PluginsIterator it = plugins_.begin(); plugins_.end() != it; ++it) {
+    res = res || it->second->IsAppForPlugin(app);
+  }
+  return res;
+}
+
+bool PluginManager::IsAppForPlugin(
+    application_manager::ApplicationSharedPtr app, ModuleID module_id) const {
+  Modules::const_iterator i = plugins_.find(module_id);
+  return i != plugins_.end() ? i->second->IsAppForPlugin(app) : false;
+}
+
+void PluginManager::OnAppHMILevelChanged(
+    application_manager::ApplicationSharedPtr app,
+    mobile_apis::HMILevel::eType old_level) {
+  DCHECK(app);
```

Comment:
@yang1070 DCHECK macro removed in release build. WE can use there DCHECK_OR_RETURN_VOID